creusot/tests/should_succeed/cc/collections.rs

```diff
@@ -0,0 +1,80 @@
+extern crate creusot_contracts;
+use creusot_contracts::*;
+use std::{
+    collections::{hash_map, HashMap, HashSet},
+    hash::Hash,
+};
+
+#[trusted]
+#[logic]
+pub fn any<T>() -> T {
+    dead
+}
+
+#[ensures(result@ == xs@)]
+pub fn roundtrip_hashmap_into_iter<K: Eq + Hash, V>(xs: HashMap<K, V>) -> HashMap<K, V> {
+    let it0 = xs.into_iter();
+    let r: HashMap<K, V> = it0.collect();
+    /*
+    let x = snapshot! { such_that(|x: (Seq<(K,V)>, &mut hash_map::IntoIter<K, V>)| {
+        let (prod, it1) = x;
+        it1.completed()
+            && it0.produces(prod, *it1)
+    })}; */
+
+    // epsilon
+    let prod = snapshot! { any::<Seq<(K,V)>>() };
+    let it1 = snapshot! { any::<&mut hash_map::IntoIter<K,V>>() };
+    proof_assert! { it1.inner().completed() };
+    proof_assert! { it0.produces(prod.inner(), *it1.inner()) };
```

Comment:
Seems like you want an `assume` here, no?

Reply:
Probably, I'm not familiar with that construct, and it doesn't seem like we expose it?

Reply:
We don't have it but it's the dual of assert, and could very easily be added. It would introduce a hypothesis to your context instead of an assertion.

Reply:
Sure, we should add it. But it should rather be called `trusted_assume`, because it is in the TCB.

Reply:
It's not solving his underlying problem but the way he's using assert looks like he meant to use an assume.

Reply:
It turns out that just asserting the existential property suffices to convince the solvers in this case.